sound/soc/sof/trace.c

```diff
@@ -30,6 +30,12 @@ static size_t sof_wait_trace_avail(struct snd_sof_dev *sdev,
 	if (sdev->host_offset > pos)
 		return sdev->host_offset - pos;
 
+	/* No data available and trace disabled -> eof */
+	if (!sdev->dtrace_is_enabled && sdev->host_offset > 0) {
```

Comment:
@kv2019i is the second check necessary? sdev->host_offset > 0?

Reply:
@ranj063 The second check is still needed as the trace client will reopen the debugfs node and we don't want to repeatedly signal eof (we do if the client repeatedly opens the file and seeks to some non-zero position and tries to read). 

Without this check, 'sof-logger -t'  will end up in a loop where it continuously opens the debugfs node, tries to read, gets eof, closes, and loop restarts.

With the check in place, sof-logger is forced to reopen the device, and its first read at position 0 will block until FW has data available. And this ensure that when the FW writes to position 0, user-space and kernel are in sync.

Reply:
@keyonjie Maybe this needs a easier to understand implementation, but in the current patch, "sdev->host_offset > 0" does have a meaning. Basicly we know "sdev->host_offset" is reset to zero at open, so this check allows us to detect the case that dtrace 1) has been been active (as it is non-zero, 2) has since been stopped (!dtrace_is_enabled), and 3) all data has been read by client (host_offset==pos). With 1+2+3 met, we can mark as eof and force the client to reopen. After it reopens, host_offset will be reset to zero and this check will no longer bet hit.

But yup, as both you and Ranjani had questions about this, I think the patch needs work to handle this case.